extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -0,0 +1,430 @@
+                requests[cacheIdx] = (request = new Request(fragmentIndex));
+                request.init();
+            }
+        } else {
+            requests[cacheIdx] = (request = new Request(fragmentIndex));
+            request.init();
+        }
+        return request;
+    }
+
+    private int cacheIndex(final long fragmentIndex) {
+        // TODO(deephaven-core#5061): Experiment with LRU caching
+        return (int) (fragmentIndex % requests.length);
+    }
+
+    private long fragmentIndex(final long pos) {
+        return pos / instructions.fragmentSize();
+    }
+
+    private String ctxStr() {
+        if (uri != null) {
+            return String.format("ctx=%d %s/%s", System.identityHashCode(S3ChannelContext.this),
+                    uri.bucket().orElseThrow(), uri.key().orElseThrow());
+        } else {
+            return String.format("ctx=%d", System.identityHashCode(S3ChannelContext.this));
+        }
+    }
+
+    final class Request
+            implements AsyncResponseTransformer<GetObjectResponse, ByteBuffer>, BiConsumer<ByteBuffer, Throwable> {
+
+        // implicitly + URI
+        private final long fragmentIndex;
+        private final long from;
+        private final long to;
+        private final Instant createdAt;
+        private Instant completedAt;
+        private final CompletableFuture<Void> released;
+        private CompletableFuture<ByteBuffer> consumerFuture;
+        private volatile CompletableFuture<ByteBuffer> producerFuture;
+        private GetObjectResponse response;
+        private int fillCount;
+        private long fillBytes;
+
+        private Request(long fragmentIndex) {
+            createdAt = Instant.now();
+            this.fragmentIndex = fragmentIndex;
+            from = fragmentIndex * instructions.fragmentSize();
+            to = Math.min(from + instructions.fragmentSize(), size) - 1;
+            released = new CompletableFuture<>();
+        }
+
+        void init() {
+            if (log.isDebugEnabled()) {
+                log.debug("send: {}", requestStr());
+            }
+            consumerFuture = client.getObject(getObjectRequest(), this);
+            consumerFuture.whenComplete(this);
+        }
+
+        public boolean isDone() {
+            return consumerFuture.isDone();
+        }
+
+        public int fill(long localPosition, ByteBuffer dest) throws IOException {
+            final int outOffset = (int) (localPosition - from);
+            final int outLength = Math.min((int) (to - localPosition + 1), dest.remaining());
+            final ByteBuffer fullFragment;
+            try {
+                fullFragment = get();
+            } catch (final InterruptedException | ExecutionException | TimeoutException | CancellationException e) {
+                throw handleS3Exception(e, String.format("fetching fragment %s", requestStr()));
+            }
+            dest.put(fullFragment.duplicate().position(outOffset).limit(outOffset + outLength));
```

Comment:
I prefer to skip this `duplicate` and just pre-record the full fragment's limit, then restore limit and position 0 after.